src/Driver/PgSQL/Connection.php

```diff
@@ -0,0 +1,144 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\DBAL\Driver\PgSQL;
+
+use Doctrine\DBAL\Driver\Connection as ConnectionInterface;
+use Doctrine\DBAL\Driver\Exception\NoIdentityValue;
+use Doctrine\DBAL\SQL\Parser;
+use PgSql\Connection as PgSqlConnection;
+use TypeError;
+
+use function assert;
+use function gettype;
+use function is_object;
+use function is_resource;
+use function pg_close;
+use function pg_escape_literal;
+use function pg_get_result;
+use function pg_last_error;
+use function pg_result_error;
+use function pg_send_prepare;
+use function pg_send_query;
+use function pg_version;
+use function sprintf;
+use function uniqid;
+
+final class Connection implements ConnectionInterface
+{
+    private Parser $parser;
+
+    /** @param PgSqlConnection|resource $connection */
+    public function __construct(private mixed $connection)
+    {
+        if (! is_resource($connection) && ! $connection instanceof PgSqlConnection) {
+            throw new TypeError(sprintf(
+                'Expected connection to be a resource or an instance of %s, got %s.',
+                PgSqlConnection::class,
+                is_object($connection) ? $connection::class : gettype($connection),
+            ));
+        }
+
+        $this->parser = new Parser(false);
+    }
+
+    public function __destruct()
+    {
+        if (isset($this->connection)) {
+            @pg_close($this->connection);
+        }
+
+        unset($this->connection);
+    }
```

Comment:
I had to insert this destructor to avoid a deadlock in our functional test suite. Backported as #5892.

Reply:
Let's wait for the necessity of `unset($this->connection);`

Reply:
Removed.